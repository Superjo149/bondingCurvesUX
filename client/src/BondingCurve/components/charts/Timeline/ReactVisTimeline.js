import { timeFormatDefaultLocale } from 'd3-time-format';
import english from 'd3-time-format/locale/en-US.json';
import moment from "moment";
import React from 'react';
import { Crosshair, FlexibleWidthXYPlot, LineSeries, MarkSeries, XAxis, GradientDefs, AreaSeries } from 'react-vis';
import "react-vis/dist/style.css";
import styles from "./timeline.module.scss";
import PropTypes from "prop-types";

// To prevent overflowing of large month names
timeFormatDefaultLocale({
    ...english,
    months: english.shortMonths
});

export default class ReactVisTimeline extends React.Component {
    state = {
        hoverValues: null
    }

    static propTypes = {
        onShowDetail: PropTypes.func,
        minDomain: PropTypes.number,
        height: PropTypes.number.isRequired,
        data: PropTypes.array.isRequired
    }

    _onMouseLeave = () => {
        const { onShowDetail } = this.props;

        this.setState({ hoverValues: null });

        if (onShowDetail) {
            onShowDetail()
        }
    };

    _onNearestX = (value) => {
        const { onShowDetail } = this.props;

        this.setState({ hoverValues: [value] });

        if (onShowDetail) {
            onShowDetail(value)
        }
    };


    render() {
        const { hoverValues } = this.state;
        const { minDomain, height, data } = this.props;

        const domain = minDomain ? [minDomain, moment().endOf('day').valueOf()] : null;

        return (
            <div className={styles.ocean_chart}>
                <FlexibleWidthXYPlot
                    margin={{ left: 0, right: 0, top: 10, bottom: 40 }}
                    xType="time"
                    animation
                    onMouseLeave={this._onMouseLeave}
                    height={height}
                    xDomain={domain}>

                    <LineSeries
                        strokeWidth={3}
                        className={styles.ocean_line}
                        style={{ strokeLinejoin: "round" }}
                        onNearestX={this._onNearestX}
                        data={data}
                    />

                    <GradientDefs>
                        <linearGradient id="oceanGradient" x1="0%" y1="0%" x2="0%" y2="100%">
                            <stop offset="0%" stopColor="#ff4092" stopOpacity={0.8} />
                            <stop offset="90%" stopColor="white" stopOpacity={0} />
                        </linearGradient>
                    </GradientDefs>

                    <AreaSeries
                        color={'url(#oceanGradient)'}
                        data={data}
                    />

                    {
                        hoverValues && (
                            <Crosshair
                                values={hoverValues}
                                className={styles.ocean_crosshair}
                            />
                        )
                    }
                    {
                        hoverValues && (
                            <LineSeries
                                strokeWidth={3}
                                className={styles.ocean_line}
                                style={{ strokeLinejoin: "round" }}
                                getX={(d) => hoverValues.x}
                            />
                        )
                    }

                    {
                        hoverValues && (
                            <XAxis tickLabelAngle={-35} />
                        )
                    }

                    {
                        hoverValues && (
                            <MarkSeries className={styles.ocean_crosshair_dot} size={4} data={hoverValues} />
                        )
                    }

                </FlexibleWidthXYPlot>
            </div>
        );
    }
}