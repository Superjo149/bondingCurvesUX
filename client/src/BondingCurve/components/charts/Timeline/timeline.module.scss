@import "../../../../assets/stylesheets/variables";

.ocean_chart{
    position:relative;
    overflow:hidden;

    .ocean_line{
        stroke:$brand-pink !important;
    }
    
    .ocean_crosshair_dot{
        :global{
            circle{
                fill: $brand-purple !important;
                stroke:$brand-white !important;
            }
        }
    }
    
    .ocean_crosshair{
    
        :global{
            .rv-crosshair__line{
                border-left: 2px dotted $brand-grey-lighter;
                background:transparent;
                opacity: .7;
    
            }
            .rv-crosshair__inner{
                display:none;
            }
        }
    }
}

.timeline_filter{
    color: $brand-grey-light;
    list-style-type: none;
    display: flex;
    margin: 0;
    padding: 0;
    font-family: $font-family-button;
    font-size: .8rem;
    
    :global{
        li{
            padding: .3rem;
            opacity:.6;
            cursor: pointer;

            &:hover{
                opacity:.8;
            }

            &.active{
                color:$brand-pink;
                opacity:1;
            }
        }
    }
}
