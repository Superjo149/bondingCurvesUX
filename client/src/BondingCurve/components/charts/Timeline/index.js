import PropTypes from 'prop-types';
import { Component } from 'react';
import React from "react";
import ReactVisTimeline from './ReactVisTimeline';
import styles from "./timeline.module.scss";
import cn from "classnames";
import moment from "moment";
import Footer from '../../Footer';

export default class Timeline extends Component {

    static propTypes = {
        bondingCurveContract: PropTypes.object.isRequired,
        web3: PropTypes.object.isRequired,
        height: PropTypes.number.isRequired,
    }

    state = {
        minDomain: 0,
        selectedItem: null,
        data: [],
        activeFilter: 'MAX',
        maxValue: 0,
        error: null
    }

    filters = ["1D", "5D", "1M", "1Y", "MAX"]

    async componentDidMount() {
        try {
            await this.getData(this.props);
        } catch (error) {
            this.setState({ error })
        }
    }

    async componentWillReceiveProps(nextProps) {
        try {
            if (nextProps.contractAddress !== this.props.contractAddress) {
                this.setState({
                    data: []
                }, async () => {
                    await this.getData(nextProps)
                })
            };
        } catch (error) {
            this.setState({ error })
        }
    }

    getData = async (props) => {
        const { bondingCurveContract, web3 } = props;

        let i = 0;

        try {

            const scale = await this.props.bondingCurveContract.methods.scale().call()

            bondingCurveContract.events.allEvents({
                fromBlock: 0,
                toBlock: 'latest'
            })
                .on('data', async (event) => {

                    if (["TokenBuyDrops", "TokenSellDrops"].indexOf(event.event) !== -1) {
                        const price = event.returnValues._price / scale;
                        const block = await web3.eth.getBlock(event.blockNumber)

                        // TODO remove subtract for release, only there to show a smoother graph for testing
                        const date = moment(block.timestamp * 1000).subtract(50 - i++, "day").valueOf();

                        let newMaxValue = this.state.maxValue;

                        if (price > this.state.maxValue) {
                            newMaxValue = price;
                        }

                        this.setState((prevState) => ({
                            data: [
                                ...prevState.data,
                                {
                                    y: +price,
                                    x: date
                                }
                            ],
                            maxValue: newMaxValue
                        }))
                    }

                })
                .on('changed', (event) => {
                    // remove event from local database
                    console.log("changed", event)
                })
                .on('error', console.error);

        } catch (error) {
            this.setState({ error })
        }
    }

    setFilter = (filter) => {
        let minDomain = 0;

        if (filter !== "MAX") {
            const duration = moment.duration(`P${filter}`)
            minDomain = moment().subtract(duration).valueOf()
        }

        this.setState({
            activeFilter: filter,
            minDomain
        })

    }

    setDetail = (selectedItem) => {
        this.setState({ selectedItem })
    }

    render() {
        const { activeFilter, selectedItem, minDomain, maxValue, data } = this.state;
        const { height } = this.props;

        if (this.state.error) throw this.state.error;

        const detail = selectedItem || data.slice(-1)[0]

        return (
            <div>
                <div>
                    <ReactVisTimeline
                        activeFilter={activeFilter}
                        minDomain={minDomain}
                        maxValue={maxValue}
                        height={height}
                        onShowDetail={this.setDetail}
                        data={this.state.data} />
                </div>

                <Footer
                    detail={detail ? {
                        title: `OCN ${detail.y.toFixed(4)}`,
                        sub: moment(detail.x).format("lll")
                    } : null}
                >
                    <ul className={styles.timeline_filter}>
                        {this.filters.map((filter) => (
                            <li
                                key={filter}
                                className={cn({ active: filter === activeFilter })}
                                onClick={this.setFilter.bind(this, filter)}>
                                {filter}
                            </li>
                        ))}
                    </ul>
                </Footer>
            </div>
        )
    }
}
