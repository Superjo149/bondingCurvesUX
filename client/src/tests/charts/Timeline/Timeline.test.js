import Enzyme, { shallow,mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import React from 'react';
import ReactDOM from 'react-dom';
import Timeline from '../../../BondingCurve/components/charts/Timeline/index';
import { bondingCurveContract, mockWeb3, mockEvents, mockEvent } from "../../mockContract";

Enzyme.configure({ adapter: new Adapter() })

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<Timeline height={200} bondingCurveContract={bondingCurveContract} web3={mockWeb3} />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('renders and should process event', () => {
  const wrapper = shallow(
    <Timeline height={200} bondingCurveContract={bondingCurveContract} web3={mockWeb3} />
  );

  wrapper.instance().handleEvent(mockEvent, .9);

  expect(mockWeb3.eth.getBlock.mock.calls.length).toBe(1)
  expect(mockWeb3.eth.getBlock).toBeCalledWith(mockEvent.blockNumber);
});

it('should filter data', () => {
  const wrapper = mount(
    <Timeline height={200} bondingCurveContract={bondingCurveContract} web3={mockWeb3} />
  );

  wrapper.instance().setFilter = jest.fn()

  expect(wrapper.find("ul li").first().hasClass('active')).toBe(false);

  wrapper.find("ul li").first().simulate('click')

  expect(wrapper.find("ul li").first().hasClass('active')).toBe(true);

  expect(wrapper.state("activeFilter")).toBe("1D");
  expect(wrapper.state("minDomain")).not.toBe(0)

});

// it('should match snapshot', () => {
//   const wrapper = shallow(
//     <Timeline height={200} bondingCurveContract={bondingCurveContract} web3={mockWeb3} />
//   );
//   expect(wrapper).toMatchSnapshot()
// });

