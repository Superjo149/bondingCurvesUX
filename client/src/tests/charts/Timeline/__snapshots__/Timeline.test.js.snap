// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Timeline
    bondingCurveContract={
      Object {
        "events": Object {
          "allEvents": [MockFunction] {
            "calls": Array [
              Array [
                Object {
                  "fromBlock": 0,
                  "toBlock": "latest",
                },
              ],
            ],
            "results": Array [
              Object {
                "isThrow": false,
                "value": EventEmitter {
                  "_events": Object {
                    "changed": [Function],
                    "data": [Function],
                    "error": [Function],
                  },
                  "_eventsCount": 3,
                  "_maxListeners": undefined,
                  "domain": null,
                },
              },
            ],
          },
        },
        "methods": Object {
          "dropsBalance": [MockFunction],
          "dropsSupply": [MockFunction],
          "ghostSupply": [MockFunction],
          "nOcean": [MockFunction],
          "ndrops": [MockFunction],
          "poolBalance": [MockFunction],
          "reserveRatio": [MockFunction],
          "scale": [MockFunction] {
            "calls": Array [
              Array [],
              Array [],
            ],
            "results": Array [
              Object {
                "isThrow": false,
                "value": Object {
                  "call": [MockFunction] {
                    "calls": Array [
                      Array [],
                      Array [],
                    ],
                    "results": Array [
                      Object {
                        "isThrow": false,
                        "value": Promise {},
                      },
                      Object {
                        "isThrow": false,
                        "value": Promise {},
                      },
                    ],
                  },
                },
              },
              Object {
                "isThrow": false,
                "value": Object {
                  "call": [MockFunction] {
                    "calls": Array [
                      Array [],
                      Array [],
                    ],
                    "results": Array [
                      Object {
                        "isThrow": false,
                        "value": Promise {},
                      },
                      Object {
                        "isThrow": false,
                        "value": Promise {},
                      },
                    ],
                  },
                },
              },
            ],
          },
          "tokenBalance": [MockFunction],
          "totalSupply": [MockFunction],
        },
      }
    }
    contractAddress="0x96eaf28b6e59defc8f736faa1681d41382d3aa32"
    height={200}
    web3={
      Object {
        "eth": Object {
          "Contract": [Function],
          "getAccounts": [MockFunction],
          "getBlock": [MockFunction],
          "getCode": [MockFunction],
          "net": Object {
            "isListening": [MockFunction],
          },
        },
        "utils": Object {
          "isAddress": [MockFunction],
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          style={
            Object {
              "minHeight": 200,
            }
          }
        >
          <ReactVisTimeline
            activeFilter="MAX"
            data={Array []}
            height={200}
            maxValue={0}
            minDomain={0}
            onShowDetail={[Function]}
          />
        </div>,
        <Footer
          detail={null}
          symbol="OCN"
        >
          <ul
            className="timeline_filter"
          >
            <li
              className=""
              onClick={[Function]}
            >
              1D
            </li>
            <li
              className=""
              onClick={[Function]}
            >
              5D
            </li>
            <li
              className=""
              onClick={[Function]}
            >
              1M
            </li>
            <li
              className=""
              onClick={[Function]}
            >
              1Y
            </li>
            <li
              className="active"
              onClick={[Function]}
            >
              MAX
            </li>
          </ul>
        </Footer>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <ReactVisTimeline
            activeFilter="MAX"
            data={Array []}
            height={200}
            maxValue={0}
            minDomain={0}
            onShowDetail={[Function]}
          />,
          "style": Object {
            "minHeight": 200,
          },
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "activeFilter": "MAX",
            "data": Array [],
            "height": 200,
            "maxValue": 0,
            "minDomain": 0,
            "onShowDetail": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <ul
            className="timeline_filter"
          >
            <li
              className=""
              onClick={[Function]}
            >
              1D
            </li>
            <li
              className=""
              onClick={[Function]}
            >
              5D
            </li>
            <li
              className=""
              onClick={[Function]}
            >
              1M
            </li>
            <li
              className=""
              onClick={[Function]}
            >
              1Y
            </li>
            <li
              className="active"
              onClick={[Function]}
            >
              MAX
            </li>
          </ul>,
          "detail": null,
          "symbol": "OCN",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <li
                className=""
                onClick={[Function]}
              >
                1D
              </li>,
              <li
                className=""
                onClick={[Function]}
              >
                5D
              </li>,
              <li
                className=""
                onClick={[Function]}
              >
                1M
              </li>,
              <li
                className=""
                onClick={[Function]}
              >
                1Y
              </li>,
              <li
                className="active"
                onClick={[Function]}
              >
                MAX
              </li>,
            ],
            "className": "timeline_filter",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "1D",
              "nodeType": "host",
              "props": Object {
                "children": "1D",
                "className": "",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "1D",
              "type": "li",
            },
            Object {
              "instance": null,
              "key": "5D",
              "nodeType": "host",
              "props": Object {
                "children": "5D",
                "className": "",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "5D",
              "type": "li",
            },
            Object {
              "instance": null,
              "key": "1M",
              "nodeType": "host",
              "props": Object {
                "children": "1M",
                "className": "",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "1M",
              "type": "li",
            },
            Object {
              "instance": null,
              "key": "1Y",
              "nodeType": "host",
              "props": Object {
                "children": "1Y",
                "className": "",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "1Y",
              "type": "li",
            },
            Object {
              "instance": null,
              "key": "MAX",
              "nodeType": "host",
              "props": Object {
                "children": "MAX",
                "className": "active",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "MAX",
              "type": "li",
            },
          ],
          "type": "ul",
        },
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            style={
              Object {
                "minHeight": 200,
              }
            }
          >
            <ReactVisTimeline
              activeFilter="MAX"
              data={Array []}
              height={200}
              maxValue={0}
              minDomain={0}
              onShowDetail={[Function]}
            />
          </div>,
          <Footer
            detail={null}
            symbol="OCN"
          >
            <ul
              className="timeline_filter"
            >
              <li
                className=""
                onClick={[Function]}
              >
                1D
              </li>
              <li
                className=""
                onClick={[Function]}
              >
                5D
              </li>
              <li
                className=""
                onClick={[Function]}
              >
                1M
              </li>
              <li
                className=""
                onClick={[Function]}
              >
                1Y
              </li>
              <li
                className="active"
                onClick={[Function]}
              >
                MAX
              </li>
            </ul>
          </Footer>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <ReactVisTimeline
              activeFilter="MAX"
              data={Array []}
              height={200}
              maxValue={0}
              minDomain={0}
              onShowDetail={[Function]}
            />,
            "style": Object {
              "minHeight": 200,
            },
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "activeFilter": "MAX",
              "data": Array [],
              "height": 200,
              "maxValue": 0,
              "minDomain": 0,
              "onShowDetail": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <ul
              className="timeline_filter"
            >
              <li
                className=""
                onClick={[Function]}
              >
                1D
              </li>
              <li
                className=""
                onClick={[Function]}
              >
                5D
              </li>
              <li
                className=""
                onClick={[Function]}
              >
                1M
              </li>
              <li
                className=""
                onClick={[Function]}
              >
                1Y
              </li>
              <li
                className="active"
                onClick={[Function]}
              >
                MAX
              </li>
            </ul>,
            "detail": null,
            "symbol": "OCN",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <li
                  className=""
                  onClick={[Function]}
                >
                  1D
                </li>,
                <li
                  className=""
                  onClick={[Function]}
                >
                  5D
                </li>,
                <li
                  className=""
                  onClick={[Function]}
                >
                  1M
                </li>,
                <li
                  className=""
                  onClick={[Function]}
                >
                  1Y
                </li>,
                <li
                  className="active"
                  onClick={[Function]}
                >
                  MAX
                </li>,
              ],
              "className": "timeline_filter",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "1D",
                "nodeType": "host",
                "props": Object {
                  "children": "1D",
                  "className": "",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "1D",
                "type": "li",
              },
              Object {
                "instance": null,
                "key": "5D",
                "nodeType": "host",
                "props": Object {
                  "children": "5D",
                  "className": "",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "5D",
                "type": "li",
              },
              Object {
                "instance": null,
                "key": "1M",
                "nodeType": "host",
                "props": Object {
                  "children": "1M",
                  "className": "",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "1M",
                "type": "li",
              },
              Object {
                "instance": null,
                "key": "1Y",
                "nodeType": "host",
                "props": Object {
                  "children": "1Y",
                  "className": "",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "1Y",
                "type": "li",
              },
              Object {
                "instance": null,
                "key": "MAX",
                "nodeType": "host",
                "props": Object {
                  "children": "MAX",
                  "className": "active",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "MAX",
                "type": "li",
              },
            ],
            "type": "ul",
          },
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
